File: app/integrations/routers/webhooks.py
from __future__ import annotations

import os
from fastapi import APIRouter, Request, HTTPException

from app.integrations.webhook_signing import verify_hmac_sha256
from app.integrations.metrics import record_request, record_latency_ms
import time


router = APIRouter(prefix="/integrations")


@router.post("/webhooks/demo")
async def demo_webhook(request: Request):

    t0 = time.time()
    record_request()

    secret = os.getenv("WEBHOOK_SECRET", "dev-secret")
    sig = request.headers.get("x-signature")

    body = await request.body()
    if not verify_hmac_sha256(secret=secret, body=body, signature=sig):
        raise HTTPException(status_code=401, detail="Invalid signature")

    # pretend to do something with payload
    # (intentionally minimal for demo)
    record_latency_ms((time.time() - t0) * 1000.0)
    return {"ok": True}

File: web/src/integrations/routes/webhooks.ts
import { Router, Request, Response } from "express";
import { verifyHmacSha256 } from "../webhookSigning";
import { recordRequest, recordLatencyMs } from "../metrics";

export const webhooksRouter = Router();

export function demoWebhook(req: Request, res: Response): void {
  const t0 = Date.now();
  recordRequest();

  const secret = process.env.WEBHOOK_SECRET || "dev-secret";
  const sig = req.headers["x-signature"] as string | undefined;
  const body = Buffer.from(JSON.stringify(req.body));

  if (!verifyHmacSha256(secret, body, sig)) {
    res.status(401).json({ error: "Invalid signature" });
    return;
  }

  recordLatencyMs(Date.now() - t0);
  res.json({ ok: true });
}

webhooksRouter.post("/integrations/webhooks/demo", demoWebhook);
